cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +344,662 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
       return Status::NotImplemented("Not implemented type kind: ", kind);
   }
 }
+}  // namespace orc
+}  // namespace adapters
+}  // namespace arrow
+
+namespace {
+
+using arrow::internal::checked_cast;
+
+arrow::Status WriteBatch(const arrow::Array& parray, int64_t orc_offset,
+                         liborc::ColumnVectorBatch* column_vector_batch,
+                         bool normalized = false);
+
+// Make sure children of StructArray have appropriate null.
+std::shared_ptr<arrow::Array> NormalizeArray(const std::shared_ptr<arrow::Array>& array) {
+  arrow::Type::type kind = array->type_id();
+  switch (kind) {
+    case arrow::Type::type::BOOL:
+    case arrow::Type::type::INT8:
+    case arrow::Type::type::INT16:
+    case arrow::Type::type::INT32:
+    case arrow::Type::type::INT64:
+    case arrow::Type::type::FLOAT:
+    case arrow::Type::type::DOUBLE:
+    case arrow::Type::type::STRING:
+    case arrow::Type::type::LARGE_STRING:
+    case arrow::Type::type::BINARY:
+    case arrow::Type::type::LARGE_BINARY:
+    case arrow::Type::type::FIXED_SIZE_BINARY:
+    case arrow::Type::type::DATE32:
+    case arrow::Type::type::DATE64:
+    case arrow::Type::type::TIMESTAMP:
+    case arrow::Type::type::DECIMAL128: {
+      return array;
+    }
+    case arrow::Type::type::STRUCT: {
+      if (array->null_count() == 0) {
+        return array;
+      } else {
+        auto struct_array = std::static_pointer_cast<arrow::StructArray>(array);
+        const std::shared_ptr<arrow::Buffer> bitmap = struct_array->null_bitmap();
+        std::shared_ptr<arrow::DataType> struct_type = struct_array->type();
+        std::size_t size = struct_type->fields().size();
+        std::vector<std::shared_ptr<arrow::Array>> new_children(size, nullptr);
+        for (std::size_t i = 0; i < size; i++) {
+          std::shared_ptr<arrow::Array> child = struct_array->field(i);
+          const std::shared_ptr<arrow::Buffer> child_bitmap = child->null_bitmap();
+          std::shared_ptr<arrow::Buffer> final_child_bitmap;
+          if (child_bitmap == nullptr) {
+            final_child_bitmap = bitmap;
+          } else {
+            final_child_bitmap = arrow::internal::BitmapAnd(
+                                     arrow::default_memory_pool(), bitmap->data(), 0,
+                                     child_bitmap->data(), 0, struct_array->length(), 0)
+                                     .ValueOrDie();
+          }
+          std::shared_ptr<arrow::ArrayData> child_array_data = child->data();
+          std::vector<std::shared_ptr<arrow::Buffer>> child_buffers =
+              child_array_data->buffers;
+          child_buffers[0] = final_child_bitmap;
+          std::shared_ptr<arrow::ArrayData> new_child_array_data = arrow::ArrayData::Make(
```

Comment:
Simpler:
```c++
auto new_child_array_data = child->data()->Copy();
new_child_array_data->buffers[0] = std::move(final_child_bitmap);
```

Reply:
Hmm this breaks the Struct of Struct test. There must be some subtle difference between the two.